src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/AnyRef.cs

```diff
@@ -3,25 +3,55 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using System.Diagnostics;
 using System.Runtime.InteropServices;
+using System.Threading;
+using Microsoft.Win32.SafeHandles;
 
 namespace System.Runtime.InteropServices.JavaScript
 {
-    public class AnyRef
+    public abstract class AnyRef : SafeHandleMinusOneIsInvalid
     {
-        public int JSHandle { get; internal set; }
-        internal GCHandle Handle;
+        public int JSHandle
+        {
+            get => (int)handle;
+        }
+
+        private AnyRef() : base(true) { }
+
+        internal GCHandle AnyRefHandle;
 
-        internal AnyRef(int jsHandle)
+        internal AnyRef(int js_handle, bool ownsHandle) : base(ownsHandle)
         {
-            JSHandle = jsHandle;
-            Handle = GCHandle.Alloc(this);
+            SetHandle((IntPtr)js_handle);
+            this.AnyRefHandle = GCHandle.Alloc(this, ownsHandle ? GCHandleType.Weak : GCHandleType.Normal);
         }
 
-        internal AnyRef(IntPtr jsHandle)
+        internal AnyRef(IntPtr js_handle, bool ownsHandle) : base(ownsHandle)
         {
-            JSHandle = (int)jsHandle;
-            Handle = GCHandle.Alloc(this);
+            SetHandle(js_handle);
+            this.AnyRefHandle = GCHandle.Alloc(this, ownsHandle ? GCHandleType.Weak : GCHandleType.Normal);
         }
+
+        // We should not provide a finalizer - SafeHandle's critical finalizer will call ReleaseHandle inside a CER for us.
+        protected override bool ReleaseHandle() => throw new NotImplementedException();
+
+#if DEBUG_HANDLE
+        private int _refCount = 0;
+
+        public void AddRef()
+        {
+            Interlocked.Increment(ref _refCount);
+        }
+
+        public void Release()
+        {
+            Interlocked.MemoryBarrier();
```

Comment:
MemoryBarrier should not be needed here. `Interlocked.Decrement` implies memory barrier.